[BUG]: add max value to IMF density. Would crash without this.

abukowski21 opened the pull request:
# [BUG] Add a cap to max level of IMF density

Weimer would crash with large values of IMF `n`. 

## Changes to GITM outputs

Hopefully none...

This will very rarely be used, but came up in the Gannon storm (May 2024)

## One more thing....

Simulating the Gannon storm meant I have to both:
- change the number of advected species
- turn off negative ion density checks

These might be something that could be added to the inputs (`#LARGEEVENT` bool flag, maybe?). 

Or, preferably I think, at least document the errors. Maybe a wiki page or a md file in `srcDoc/` with FAQ and/or common errors? Definitely prefer the second option, but open to feedback. 

Can delay approving this until that's done, if we want to do either.

Commit:
[BUG]: add max value to IMF density. Would crash without this.

aaronjridley commented:
Ok - I think that we decided that we should check on the 45 deg lat issue and see if that changes anything (which we suspect won't make a difference in normal runs).  Then return to this.

zihan-wang-space commented:
In GITM Ualibrary, Weimer potential is set to zero beyond 45 latitude. During the Gannon storm, the potential can be beyond 45 latitude, generating a large electric field at 45 latitude.

aaronjridley commented:
Yeah, Aaron B figured out where you were talking about and we discussed this earlier today.  I think that changing this to 10 deg makes a lot of sense and should be fine. We then probably wouldn't need the density cap, but he is going to check whether we need both or not.  Thanks for finding this!

abukowski21 commented:
Hey all, finally had time to look into this... I think it's going to be worthwhile to have both fixes, so thank you for pointing it out, Zihan.

I was experimenting with the indices and found quite a few configurations where Weimer will error. So I think that, moving forward, we'll need to check particularly intense storms and adjust the maximum value of the indices if things end up crashing in Weimer.

Quick table of current limits & an example where I got Weimer to crash:

|   Index       |      Limit |  Crash |
|   ----       |      ---- |  ----  |
|  $B_z$   |   $\pm 50 $  |  -30  |
|  $B_y$           | $\pm 50 $  |  -5  |
|  $V_{total}$  | $\pm 2000$  |  1200  |
|  $n$              |  $\pm 80 $ (new) |  70  |

So it may not be as simple as capping certain indices, since the lower latitude limit will depend on all of them.  For example, changing each index in the `Crash` case, one at a time, ($B_z$ -30 -> -10, $V_{total}$ 1200 -> 800, $n$ 70 -> 10) got the run to work. This was all with the latitude limit set to 10 degrees. I think this shows we need to go on a storm-by-storm basis and just change the limits if one index goes particularly high. Since I'm not aware of Weimer crashing in any other storms, it's probably okay to leave the fix for the Gannon storm in now and then update the limits after the next huge storm.


But the adjusting the latitude limit definitely improved stability, so thanks for finding it! I'll get it added to the [Electrodynamics repository](https://github.com/GITMCode/Electrodynamics).

abukowski21 committed:
Do not stop on encountering negative ion densities

abukowski21 committed:
Update ModEarth.f90 to only advect one ion species

Aaron R, is this what we wanted to do until we can look more into it? Can revert if not.

abukowski21 commented:
Added the other two fixes necessary to get the Gannon storm finished.  I think we discussed adding these to the develop branch, but we can hold off till we know exactly how things will be changed by only advecting one ion species. It's easy to revert that commit.

This is ready for review, @aaronjridley

aaronjridley reviewed (commented):
good. merge...
